File: src/net/merayen/elastic/ui/controller/Controller.java
package net.merayen.elastic.ui.controller;

import org.json.simple.JSONObject;

import net.merayen.elastic.ui.objects.top.Top;
import net.merayen.elastic.util.Postmaster;

public abstract class Controller {
	protected Gate gate;

	public Controller(Gate gate) {
		this.gate = gate;
	}

	protected abstract void onInit();

	/**
	 * Message received from the backend.
	 */
	protected abstract void onMessageFromBackend(Postmaster.Message message);

	/**
	 * Message sent from the UI.
	 */
	protected abstract void onMessageFromUI(Postmaster.Message message);

	public void sendToBackend(Postmaster.Message message) {
		gate.to_backend.send(message);
	}

	public Top getTopObject() {
		return gate.top;
	}

	protected abstract JSONObject onDump();
	protected abstract void onRestore(JSONObject obj);
}

File: src/main/kotlin/net/merayen/elastic/ui/controller/Controller.kt
package net.merayen.elastic.ui.controller

import net.merayen.elastic.system.intercom.ElasticMessage
import net.merayen.elastic.ui.objects.top.Top
import net.merayen.elastic.ui.objects.top.views.View
import net.merayen.elastic.util.NetListMessages
import net.merayen.elastic.util.Postmaster
import java.util.*

abstract class Controller(val top: Top) {
	private val messagesToBackend = Postmaster<ElasticMessage>()

	abstract fun onInit()

	/**
	 * Message received from the backend.
	 */
	abstract fun onMessageFromBackend(message: ElasticMessage)

	/**
	 * Message sent from the UI.
	 */
	abstract fun onMessageFromUI(message: ElasticMessage)

	fun sendToBackend(message: ElasticMessage) = messagesToBackend.send(message)

	fun retrieveMessagesToBackend(): Collection<ElasticMessage> {
		return messagesToBackend.receiveAll()
	}

	protected fun <T : View> getViews(cls: Class<T>): List<T> {
		val result = ArrayList<T>()

		for (w in top.getWindows())
			if (w.isInitialized)
				for (vp in w.viewportContainer.viewports)
					if (vp.view.javaClass.isAssignableFrom(cls))
						result.add(vp.view as T)

		return result
	}

	/**
	 * Use by NodeView to restore itself from the current NetList.
	 */
	fun getNetListRefreshMessages() = NetListMessages.disassemble(top.netlist)
}
